kong/db/dao/plugins/go.lua

```diff
@@ -79,7 +121,7 @@ do
   end
 
   pcall(ffi.metatype, ffi_sock, {
```

Comment:
Actually some second thought on this, is it possible to fallback to LuaSocket instead of doing raw BSD socket calls like the rest of the code base?

Reply:
AFAICT, luasocket doesn't have the equivalent of the receiveany operation, which is important for the protocol

Reply:
Is it possible to do the framing ourself (e.g. send the length of the msgpack payload before the payload) instead of relying on this behavior?

Reason being, using `receiveany` is not guaranteed to work anyway, since the stream socket never guarantees message boundary, that's an upper layer protocol concern.

Reply:
The decoder calls the receive function until it can decode a whole object.  Reading any fixed size over one byte at a time could block indefinitely. Receiveany returns what is available.  Can be less than what's sent, so it's repeated.

Reply:
The good news is that initialization is done via pipes and we're not supporting filter phases, so this ffi option might be removed eventually

Reply:
Yes I think removing it will be the best. Still I don’t think the way we are using `receiveany` is correct. The unpacker could possibly been reading more data than it needs, and because we don’t keep the unpacker around, subsequent reads from the socket could return partially read data.    

I’m definitely not referring to making 1 byte reads for all the data, that would be extremely inefficient. What I proposed is to send a 2 byte length before each msgpack message, on the other end read the fixed 2 byte length first then use its value to read exactly the needed bytes from the socket. 

Also from the look of the code, the plugin server may spontaneously send data to Kong even if Kong didn’t ask for it? If that is the case this may need to be changed, as a keepalived connection will immediately error out if Nginx receives any data on it. The only safe interaction In this case is a request-response flow.

Reply:
I think another possible solution is we can simply use the data gram mode Unix socket, that way keep alive or framing are no concern.

Reply:
Or is the plugin server only gonna respond one or more msgpack payloads when the Kong worker sends the call? If that is the case I think `receiveany` streaming read would work. Sorry for the extra comments but it was not clear at first look and I want to make sure we are using `receiveany` correctly here :)